core/include/traccc/fitting/kalman_filter/kalman_actor.hpp

```diff
@@ -33,31 +34,50 @@ struct kalman_actor : detray::actor {
         state(vector_t<track_state_type>&& track_states)
             : m_track_states(std::move(track_states)) {
             m_it = m_track_states.begin();
+            m_it_rev = m_track_states.rbegin();
```

Comment:
Does it have to be a reverse iterator? Is it not possible to set a bidirectionl iterator to the last element (```m_itr = detray::detail::prev(m_track_states.end())```) and go back? Then it would not need a second iterator and the direction can be reversed at any time. This could be put at a single point in :
```
void reset() {
     m_it = backward_mode ? detray::detail::prev(m_track_states.end()) : m_track_states.begin();
 }
```

Also, if there were two different methods ```next()``` and ```previous()```, they could be called in the if-else block in line 141 and there wound not need to be an if-else branch in ```next()```

Reply:
Otherwise, I am not so sure why these two workflows are contained in a single class? They seem to be pretty separate from each other, except that they act on the same data. We continue to switch between them with a boolean flag at every step of the algorithm? There are two different actor chains for fitting and smoothing anyway...

Reply:
> I think it is convenient to have them in the same actor because smoother need the filtered state information from the forward filter.

Does the smoothing step need to own the track states? It also seems a bit inconvenient to always have to add an ```if(backward_mode) {...} else {...}``` everywhere. But I leave that up to you

> Using `detail::prev(m_track_states.end())` sounds interesting but I am not very familiar with this grammar. I don't think it is going to make big difference either

It is the same as ```std::prev```, just working on device. It moves an iterator one position backwards, since ```end()``` is one position behind the last element of the range

Reply:
> Does the smoothing step need to own the track states? 

As long I don't want to copy the forward filtered state to somewhere else, yes it need to own

Reply:
> It is the same as std::prev, just working on device.

Anyway you will still need to have if-block for `next` or `is_complete`. (Not even sure how I can write `is_complete` function with `std::prev`)

Reply:
> > It is the same as std::prev, just working on device.
> 
> Anyway you will still need to have if-block for `next` or `is_complete`. (Not even sure how I can write `is_complete` function with `std::prev`)

My initial point was that if you call ```next()``` in the if block in l.141 and a new method ```previous()``` in the else block below, the if-else block in ```next()``` would not actually be needed... The ```operator()``` would also not need an if-else block anymore.

Yes, ```is_complete()``` would need to be checked against ```begin()```, that is true

Reply:
> I am not following, it just dereferences an iterator to get a reference to the track state or not? It can do that without owning the track_state container that the iterator points to or what am I missing?

You are right. So is your suggestion to separate the `track_states` from the actor and have it accessed by two different actors made for forward filter and backward filter?

Reply:
Someone might attempt that in the next PRs but let's keep the changes minimal in this PR.

Reply:
> > I am not following, it just dereferences an iterator to get a reference to the track state or not? It can do that without owning the track_state container that the iterator points to or what am I missing?
> 
> You are right. So is your suggestion to separate the `track_states` from the actor and have it accessed by two different actors made for forward filter and backward filter?

Yes, instead of moving the ```input_states``` into the fitter state, the fitter/smoother states could simply be views onto the track states (views in the sense of std::ranges not vecmem)